### DISCUSSION TITLE
Example code to run a simple shader/program on the gpu with Sokol?

### DISCUSSION THREAD

**Discussion by l1mey112** [vlang/v#14036] [General]
I'm semi new to opengl and sokol in general. I want to find a way to write in GLSL and execute it on the gpu, hopefully in a parallelised way. Without creating vertex buffers or messing with windows, just a simple set of code with an input and an output ran on the gpu.

Thank you.

**Comment by larpon** [comment]
The ["simple shader glsl" example](https://github.com/vlang/v/tree/44d2647e2d41e54fc6543d9aee6856893957c89c/examples/sokol/07_simple_shader_glsl) is currently the simplest form of shader example we have - there are more advanced examples in the `examples/sokol/` directory also.

If you're in V's project root dir you could do:
```
cd examples/sokol/simple_shader_glsl
v shader .
v run .
```
You'll have to run `v shader .` everytime you change the code in the shader (`simple_shader.glsl`)
You can read more about the shader tools and format [here](https://github.com/floooh/sokol-tools/blob/master/docs/sokol-shdc.md)

(`v shader` is just a convenience wrapper around the `sokol-shdc` tool).

That's as simple as we can get it at this stage - hope it helps

**Reply by l1mey112** [reply]
Is there a way to write an execute code similar to a general purpose cuda program?

**Reply by larpon** [reply]
Not that I'm aware of, no. The shader ecosystem is heavily tied to graphical use only.

**Reply by larpon** [reply]
Maybe you can find a C library that does?
V is very easy to do C interop with.

**Reply by l1mey112** [reply]
I may try using cuda with C then use with V

**Reply by larpon** [reply]
That's probably the best bet - until we have something more integrated up and running 👍

**Reply by snakeru** [reply]
@l1mey112 did you find a way? I also need to get the compute shader results back in the CPU and it seems not possible at the moment (though there was some preliminary compute shader support merged to sokol 1-2 months ago, but not enough for this).

This is definitely not helped by the fact that this is the first real shader that I'm intending to write.

**Reply by Eliyaan** [reply]
Maybe take a look at https://github.com/vlang/vsl/tree/main/vcl

**Reply by snakeru** [reply]
@Eliyaan yes, that one looks extremely promising. I spent a couple of hours yesterday trying to get it running, but didn't succeed, it was complaining about inability to find the `get_devices()` function.
However, today I tried it again and it finally worked. The solution was to 

    cd .vmodules ; ln -s vsl/vcl .

I don't quite understand, why, but so long it works, I'm happy. Thank you for making me try it again!
